Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='green'><b>Полезные (и просто интересные) материалы:</b></font> \
Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 

<font color='green'>Пример BERT с GPU:
```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)

Если понравилась работа с текстами, то можешь посмотреть очень интересный (но очень-очень сложный) курс лекций: https://github.com/yandexdataschool/nlp_course .
</font>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Радует, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* Отлично, что стоп-слова были исключены при векторизации!
* Над этим проектом нужно будет еще немного поработать. Однако, изменения не должны занять много времени.
* С радостью отвечу на твои вопросы, если они есть. Лучше всего их собрать в следующей ячейке. Жду новую версию проекта.


<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> 
   
Добрый день, Артём!
Благодарю за подсказки, я постарался их максимально использовать.

    
Какие правки мною внесены:
    
    
    пункт 2.3, не использовал перевод в юникод
    
    пункт 2.3, - поменял лемматизатор на правильный
    
    пункт 3 - при правильной лемматизации качество моделей изменилось, пришлось дообучить еще 2, из которых оптимальной стала SVC()
    
    пункт 4 - провёл эксперимент с пайплайном. Целиком проект не переписал, тк в настоящее время не располагаю приблизительно 4 часами на кросс-валидацию
    
    пункт 5 и 6 - поменял в соответствии с изменившимися результатами работы
 
</div>



### <font color='orange'>Общее впечатление (ревью 2)</font>
* После исправлений проект улучшился и теперь он может быть зачтен.
* Для удобства все новые комментарии обозначены фразой "ревью 2".
* Удачи в дальнейшем обучении и следующих работах!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li><li><span><a href="#Общее-впечатление-(ревью-2)" data-toc-modified-id="Общее-впечатление-(ревью-2)-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span><font color="orange">Общее впечатление (ревью 2)</font></a></span></li></ul></li><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Цель-и-задачи-работы." data-toc-modified-id="Цель-и-задачи-работы.-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Цель и задачи работы.</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Загрузка-и-осмотр-датасета" data-toc-modified-id="Загрузка-и-осмотр-датасета-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Загрузка и осмотр датасета</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Расчет-TF-IDF" data-toc-modified-id="Расчет-TF-IDF-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Расчет TF-IDF</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#LogisticRegression()" data-toc-modified-id="LogisticRegression()-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>LogisticRegression()</a></span></li><li><span><a href="#RandomForestClassifier()" data-toc-modified-id="RandomForestClassifier()-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>RandomForestClassifier()</a></span></li><li><span><a href="#LGBMClassifier" data-toc-modified-id="LGBMClassifier-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>LGBMClassifier</a></span></li><li><span><a href="#SVC" data-toc-modified-id="SVC-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>SVC</a></span></li><li><span><a href="#LinearSVC" data-toc-modified-id="LinearSVC-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>LinearSVC</a></span></li></ul></li><li><span><a href="#Pipeline" data-toc-modified-id="Pipeline-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Pipeline</a></span></li><li><span><a href="#Тестирование-лучшей-модели" data-toc-modified-id="Тестирование-лучшей-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Тестирование лучшей модели</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Выводы</a></span><ul class="toc-item"><li><span><a href="#Цель-и-задачи-работы." data-toc-modified-id="Цель-и-задачи-работы.-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Цель и задачи работы.</a></span></li><li><span><a href="#Материалы-и-методы" data-toc-modified-id="Материалы-и-методы-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Материалы и методы</a></span></li><li><span><a href="#Ход-выполнения-проекта" data-toc-modified-id="Ход-выполнения-проекта-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Ход выполнения проекта</a></span></li><li><span><a href="#Результаты-исследования" data-toc-modified-id="Результаты-исследования-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Результаты исследования</a></span></li><li><span><a href="#Практические-рекомендации" data-toc-modified-id="Практические-рекомендации-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Практические рекомендации</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»
## Описание проекта

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

### Цель и задачи работы.
- Цель: обучить модель классифицировать комментарии на позитивные и негативные со значением метрики качества F1 не меньше 0.75.

- Задачи:
    - Очистить и лемматизировать корпус;
    - Рассчитать векторное представление текста на основе TF-IDF;
    - Получить обучающую и тестовую выборки, выделить признаки и целевой признак;
    - Обучить различные модели и проанализировать гиперпараметры;
    - Выбрать и протестировать оптимальную модель.

### Описание данных

Работа будет выполнена на размеченном датасете с комментариями пользователей, полученном от заказчика. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

### Импорт библиотек

In [33]:
import pandas as pd
import numpy as np
import re

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score
from sklearn.svm import SVC, LinearSVC
from sklearn.pipeline import Pipeline

from lightgbm import LGBMClassifier



In [2]:
import nltk
from nltk.corpus import stopwords as nltk_stopwords
nltk.download('stopwords')
from nltk.stem import WordNetLemmatizer

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в верхних ячейках ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

### Загрузка и осмотр датасета

In [3]:
data = pd.read_csv('/datasets/toxic_comments.csv')

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


In [5]:
data.head()

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

In [6]:
data = data.drop('Unnamed: 0', axis=1)

In [7]:
print("Количество пропусков в данных:", data.isna().sum())

Количество пропусков в данных: text     0
toxic    0
dtype: int64


In [8]:
print("Количество дубликатов в данных:", data.duplicated().sum())

Количество дубликатов в данных: 0


In [9]:
print("Баланс целевого признака: доля положительного класса составил", data['toxic'].mean())

Баланс целевого признака: доля положительного класса составил 0.10161213369158527


При первичном исследовании данных найден неинформативный столбец 'Unnamed: 0', который удалён.
Пропусков в данных нет.
Дубликатов нет.
Определяется значительный дисбаланс классов целевого признака, доля положительного класса приблизительно 10%.

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корреткно. Радует, что баланс классов был изучен.
</div>

### Лемматизация
Ниже лемматизируем текст и удалим символы, не используемые при расчёте TF-IDF.

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Приводить тексты к юникоду не имеет смысла, так как они все на английском. Это может привести к падению ядра из-за увеличения объема занимаемой памяти.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента</b> 
    
Спасибо! убрал
</div>

<div class="alert alert-block alert-success">
<b>Успех (ревью 2):</b> ОК.
</div>

In [10]:
corpus = data['text']
corpus

0         Explanation\nWhy the edits made under my usern...
1         D'aww! He matches this background colour I'm s...
2         Hey man, I'm really not trying to edit war. It...
3         "\nMore\nI can't make any real suggestions on ...
4         You, sir, are my hero. Any chance you remember...
                                ...                        
159287    ":::::And for the second time of asking, when ...
159288    You should be ashamed of yourself \n\nThat is ...
159289    Spitzer \n\nUmm, theres no actual article for ...
159290    And it looks like it was actually you who put ...
159291    "\nAnd ... I really don't think you understand...
Name: text, Length: 159292, dtype: object

In [11]:
corpus = corpus.apply(lambda x:re.sub(r'[^a-zA-Z]', ' ', x))
# с помощью регулярного выражения оставляем только символы с a по z,
# а также A по Z, проче заменяем на пробелы 

<div class="alert alert-block alert-success">
<b>Успех:</b> Очистка была сделана верно.
</div>

In [12]:
corpus = corpus.apply(lambda x: x.split())

# удаляем образовавшиеся в процессе замены "лишние" пробелы

In [13]:
corpus

0         [Explanation, Why, the, edits, made, under, my...
1         [D, aww, He, matches, this, background, colour...
2         [Hey, man, I, m, really, not, trying, to, edit...
3         [More, I, can, t, make, any, real, suggestions...
4         [You, sir, are, my, hero, Any, chance, you, re...
                                ...                        
159287    [And, for, the, second, time, of, asking, when...
159288    [You, should, be, ashamed, of, yourself, That,...
159289    [Spitzer, Umm, theres, no, actual, article, fo...
159290    [And, it, looks, like, it, was, actually, you,...
159291    [And, I, really, don, t, think, you, understan...
Name: text, Length: 159292, dtype: object

In [14]:
 
wnl = WordNetLemmatizer()
# объявляем объект WordNetLemmatizer
 
 

In [15]:
def lemmatize_list(list):
    for i in range(len(list)):
        list[i] = wnl.lemmatize(list[i])
    return list

In [16]:
%%time
corpus = corpus.apply(lemmatize_list)

CPU times: user 32.6 s, sys: 141 ms, total: 32.7 s
Wall time: 33.7 s


In [17]:
corpus = corpus.apply(lambda x: " ".join(x))
# объединяем леммы

In [18]:
corpus

0         Explanation Why the edits made under my userna...
1         D aww He match this background colour I m seem...
2         Hey man I m really not trying to edit war It s...
3         More I can t make any real suggestion on impro...
4         You sir are my hero Any chance you remember wh...
                                ...                        
159287    And for the second time of asking when your vi...
159288    You should be ashamed of yourself That is a ho...
159289    Spitzer Umm there no actual article for prosti...
159290    And it look like it wa actually you who put on...
159291    And I really don t think you understand I came...
Name: text, Length: 159292, dtype: object

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Этот лемматизатор работает только с русским языком, так что его нет смысла применять. Я бы использовал WordNetLemmatizer. Отмечу, что его нужно применять к словам, а не ко всему тексту сразу.
</div>

<div class="alert alert-block alert-warning">
<b>Ошибка:</b> Благодарю. Исправлено!
</div>

<div class="alert alert-block alert-success">
<b>Успех (ревью 2):</b> Отлично!
</div>

Корпус лемматизирован. 

В качестве признаков при ML будет использован показатель TF-IDF.

### Расчет TF-IDF

In [19]:
%%time
stopwords = set(nltk_stopwords.words('english'))
# задаем стоп-слова
vect = TfidfVectorizer(stop_words=stopwords)
# создаем объект класса TfidfVectorizer
features_train, features_test, target_train, target_test = train_test_split(
    corpus, data['toxic'], test_size = 0.25, stratify=data['toxic'], random_state=13)
# создаем стратифицированные выборки - обучающую и тестовую

features_train = vect.fit_transform(features_train)
# обучаем TfIdfVectorizer только на трэйне

features_test = vect.transform(features_test)
# применяем также к тестовой выборке

for _ in [features_train, features_test, target_train, target_test]:
    print(_.shape)
# убедимся, что размеры выборок соответствуют ожиданиям

(119469, 138039)
(39823, 138039)
(119469,)
(39823,)
CPU times: user 7.77 s, sys: 98.1 ms, total: 7.87 s
Wall time: 8.93 s


TF-IDF рассчитана, выборки получены. Можно приступать к обучению. В данной работе будут использованы модели логистической регрессии, случайного леса и классификтор LGBM.


<div class="alert alert-block alert-success">
<b>Успех:</b> Разбиение было сделано верно. Отлично, что векторизатор был обучен только на тренировочной части данных.
</div>

## Обучение

### LogisticRegression()

In [ ]:
# param_grid_lin={'solver': ['liblinear', 'lbfgs', 'sag', 'saga', 'newton-cg'],
#                 'class_weight': ['balanced'],
#                 'C':[0.6, 0.7,0.8, 0.9, 1]
#                 'max_iter':[50, 100, 200]
#                 'random_state':[13]}

# При подборе гиперпараметров использовался данный parameter grid. Он закомментирован и заменён оптимальным набором
# с целью ускорения работы кода.

In [ ]:
param_grid_lin={'solver': ['lbfgs','sag', 'saga', 'newton-cg'],
                'class_weight': ['balanced'],
                'max_iter':[300],
                'C':[1, 0.9],
                'random_state':[13]}

In [ ]:
# scaler = StandardScaler(with_mean=False)
# features_scal_train = scaler.fit_transform(features_train)

# Скейлер при попытке его использования значимо ухудшил
# результаты, поэтому он не применен в окончательной версии. 

In [ ]:

linear_cv = GridSearchCV(estimator=LogisticRegression(),
                      param_grid=param_grid_lin,
                      scoring='f1',
                      n_jobs=-1,
                         
                         
                      cv=3,
                      verbose=2)


In [ ]:
%%time
linear_cv.fit(features_train, target_train)
print('Лучший результат F1: {:.2f}'.format(linear_cv.best_score_))
print('при оптимальных гиперпараметрах:\n', linear_cv.best_params_)

Модель логистической регрессии позволила добиться значения F1: 0.74. 

при гиперпараметрах:  {'C': 1, 'class_weight': 'balanced', 'max_iter': 150, 'random_state': 13, 'solver': 'liblinear'}

Данное значение F1 не удовлетворяет требованиям заказчика, поэтому модель использоваться в дальнейшем не будет.


<div class="alert alert-block alert-info">
<b>Совет: </b> Напомню, что внутри кросс-валидации происходит разбиение выборки на треин и валидацию. Однако, в таком случае векторизатор обучен на всей выборке, а это не совсем корректно. Для избежания такого эффекта можно использовать <a href="https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html">пайплайн</a>. <a href="https://medium.com/analytics-vidhya/ml-pipelines-using-scikit-learn-and-gridsearchcv-fe605a7f9e05">Тут</a> есть пример.
</div>


<div class="alert alert-block alert-warning">
<b>Комментарий студента: </b>
    
Спасибо, провёл эксперимент с пайплайном, выделив его в отдельный пункт 4. Как работает, разобрался, однако если переписать под него весь проект, на кроссвалидацию всех моделей уйдёт не менее 4 часов, а я сейчас не располагаю таким временем =( поэтому не переписал проект целиком.
    
</div>



### RandomForestClassifier()

In [ ]:
# param_grid_forest = {'random_state': [13],
#                      'class_weight': ['balanced'],
#                      'max_depth': range(1,26),
#                      'n_estimators': [20, 50, 80, 100, 150, 190, 200]}

# При подборе гиперпараметров использовался данный parameter grid. Он закомментирован и заменён оптимальным набором
# с целью ускорения работы кода.

In [ ]:
param_grid_forest = {'random_state': [13],
                     'class_weight':['balanced'],
                     'max_depth': [25],
                     'n_estimators': [190]}

In [ ]:
%%time
forest_cv = GridSearchCV(estimator=RandomForestClassifier(),
                      param_grid=param_grid_forest,
                      scoring='f1',
                      n_jobs=-1,
                      cv=3,
                      verbose=2)
forest_cv.fit(features_train, target_train)
print(forest_cv.best_score_)
print(forest_cv.best_params_)

Модель случайного леса позволила добиться значения F1: 0,42

при гиперпараметрах: {'class_weight': 'balanced', 'max_depth': 25, 'n_estimators': 190, 'random_state': 13}

Данная модель не достигла уровня F1-метрики, требуемого заказчиком, поэтому в дальнейшем использоваться не будет.


### LGBMClassifier

In [ ]:
# param_grid_lgbm = {'random_state':[13],
                  
#                   'num_leaves': range(10,41,10),
#                   'n_estimators': [30, 50, 100, 150, 300, 450],
#                   'max_depth': range(1, 16),
#                   'learning_rate': [1, 0.1],
#                   'class_weight':['balanced']}
# Этот parameter grid использовался при подборе

In [ ]:
# Здесь указаны лучшие значения для быстродействия кода

param_grid_lgbm = {'random_state':[13],
                  
                  'num_leaves': [25],
                  'n_estimators': [440],
                  'max_depth': [9],
                  'learning_rate': [1],
                  'class_weight':['balanced'],
                  #'objective':['binary']
                  }

In [ ]:

lgbm_cv = GridSearchCV(estimator=LGBMClassifier(),
                      param_grid=param_grid_lgbm,
                      scoring='f1',
                      n_jobs=-1,
                      cv=3,
                      verbose=2)


In [ ]:
%%time
lgbm_cv.fit(features_train, target_train)
print(lgbm_cv.best_score_)
print(lgbm_cv.best_params_)

При подборе гиперпараметров LGBMClassifier не удалось достичь требуемых значений F1-метрики.
- Максимальное полученное значение F1: 0.74
- при гиперпараметрах 'class_weight': 'balanced', 'learning_rate': 1, 'max_depth': 9, 'n_estimators': 440, 'num_leaves': 25



### SVC

In [20]:
param_grid_svc = {'random_state':[13],
                  'class_weight':['balanced'],
                  'verbose':[True]
                  }

In [22]:
svc_cv = GridSearchCV(estimator=SVC(),
                      param_grid=param_grid_svc,
                      scoring='f1',
                      n_jobs=-1,
                      cv=2,
                      verbose=2)

In [23]:
%%time
svc_cv.fit(features_train, target_train)
print(svc_cv.best_score_)
print(svc_cv.best_params_)

Fitting 2 folds for each of 1 candidates, totalling 2 fits
[LibSVM]............................*.........*
optimization finished, #iter = 37931
obj = -6000.906167, rho = -0.610160
nSV = 24421, nBSV = 5979
Total nSV = 24421
[CV] END class_weight=balanced, random_state=13, verbose=True; total time=28.9min
[LibSVM]............................*..........*
optimization finished, #iter = 38957
obj = -6020.368974, rho = -1.621700
nSV = 24979, nBSV = 5914
Total nSV = 24979
[CV] END class_weight=balanced, random_state=13, verbose=True; total time=29.2min
[LibSVM]................................................*.................*
optimization finished, #iter = 65353
obj = -11023.936380, rho = -0.938072
nSV = 41732, nBSV = 11429
Total nSV = 41732
0.7538032387677578
{'class_weight': 'balanced', 'random_state': 13, 'verbose': True}
CPU times: user 2h 13min 27s, sys: 20.7 s, total: 2h 13min 48s
Wall time: 2h 13min 58s


In [24]:
%%time
f1_score(target_test, svc_cv.predict(features_test))

0.7802153968887116

При кросс-валидации модели SVC получено F1-значение 0.78.

### LinearSVC

In [28]:
param_grid_linsvc = {'random_state':[13],
                  'class_weight':['balanced'],
                  'verbose':[True]
                  }

In [29]:
linsvc_cv = GridSearchCV(estimator=LinearSVC(),
                      param_grid=param_grid_linsvc,
                      scoring='f1',
                      n_jobs=-1,
                      cv=2,
                      verbose=2)

In [30]:
linsvc_cv.fit(features_train, target_train)
print(linsvc_cv.best_score_)
print(linsvc_cv.best_params_)

Fitting 2 folds for each of 1 candidates, totalling 2 fits
[LibLinear]..............*.
optimization finished, #iter = 150
Objective value = -6040.450386
nSV = 21063
[CV] END class_weight=balanced, random_state=13, verbose=True; total time=   1.0s
[LibLinear].................*
optimization finished, #iter = 173
Objective value = -6018.078676
nSV = 21367
[CV] END class_weight=balanced, random_state=13, verbose=True; total time=   1.0s
[LibLinear]..................*
optimization finished, #iter = 186
Objective value = -13027.893975
nSV = 38742
0.7515088622701257
{'class_weight': 'balanced', 'random_state': 13, 'verbose': True}


Модель LinearSVC() при кросс-валидации показала значение F1 0.75.

Модель SVC() ввиду наивысшего результата F1 признана лучшей.

## Pipeline
 С целью элиминации target leakage при кросс-валидации используем pipeline.

In [34]:
features_pipe_train, features_pipe_test, target_pipe_train, target_pipe_test = train_test_split(
    corpus,
    data['toxic'],
    test_size=0.25,
    random_state=13,
    stratify=data['toxic'])

In [45]:
linsvc_pipe_grid = param_grid_linsvc = [{'clf__random_state':[13],
                  'clf__class_weight':['balanced'],
                  'clf__verbose':[True]
                  }]

In [46]:
linsvc_pipe = Pipeline([('preprocess', TfidfVectorizer(stop_words=stopwords)),
 ('clf', LinearSVC())])

In [47]:
linsvc_pipe_cv = GridSearchCV(estimator=linsvc_pipe,
                      param_grid=linsvc_pipe_grid,
                      scoring='f1',
                      n_jobs=-1,
                      cv=2,
                      verbose=2)

In [48]:
linsvc_pipe_cv.fit(features_pipe_train, target_pipe_train)

Fitting 2 folds for each of 1 candidates, totalling 2 fits
[LibLinear]..............*
optimization finished, #iter = 148
Objective value = -6075.690248
nSV = 20983
[CV] END clf__class_weight=balanced, clf__random_state=13, clf__verbose=True; total time=   7.1s
[LibLinear].................*
optimization finished, #iter = 174
Objective value = -6054.946219
nSV = 21270
[CV] END clf__class_weight=balanced, clf__random_state=13, clf__verbose=True; total time=   7.2s
[LibLinear]..................*
optimization finished, #iter = 186
Objective value = -13027.893975
nSV = 38742


GridSearchCV(cv=2,
             estimator=Pipeline(steps=[('preprocess',
                                        TfidfVectorizer(stop_words={'a',
                                                                    'about',
                                                                    'above',
                                                                    'after',
                                                                    'again',
                                                                    'against',
                                                                    'ain',
                                                                    'all', 'am',
                                                                    'an', 'and',
                                                                    'any',
                                                                    'are',
                                                                    'aren',
    

In [49]:
linsvc_pipe_cv.best_score_

0.749127876539166

In [50]:
f1_score(target_pipe_test, linsvc_pipe_cv.predict(features_pipe_test))

0.759327498848457

<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что попробовала разные модели в этом шаге!
</div>


<div class="alert alert-block alert-success">
<b>Успех (ревью 2):</b> С пайплайном все ОК!
</div>

## Тестирование лучшей модели

Лучшего результата F1 при кросс-валидации удалось достичь с помощью модели SVC.

In [ ]:
print('Значение F1 лучшей модели на тестовой выборке: {:.2f}'.format(
f1_score(target_test, svc_cv.predict(features_test))))

Результаты тестирования SVC: F1 = 0.78, что соответствует требованиям заказчика.

Данное значение метрики получено при использовании гиперпараметра class_weight='balanced', ядро 'rbf', количество итераций не ограничено.

In [31]:
print('Значение F1 лучшей модели на тестовой выборке: {:.2f}'.format(
f1_score(target_test, linsvc_cv.predict(features_test))))

Значение F1 лучшей модели на тестовой выборке: 0.76


При тестировании SVC(), обученной на всей трэйн-выборке, получено F1-значение 0.76, что соответствует требованиям к работе. Данную модель можно рекомендовать для практического применения при заданном значении F1.

<div class="alert alert-block alert-success">
<b>Успех:</b> Тестирование было сделано правильно.
</div>

## Выводы
### Цель и задачи работы.
- Цель: обучить модель классифицировать комментарии на позитивные и негативные со значением метрики качества F1 не меньше 0.75.

- Задачи:
    - Очистить и лемматизировать корпус;
    - Рассчитать векторное представление текста на основе TF-IDF;
    - Получить обучающую и тестовую выборки, выделить признаки и целевой признак;
    - Обучить различные модели и проанализировать гиперпараметры;
    - Выбрать и протестировать оптимальную модель.
    
### Материалы и методы

Работа выполнена на датасете, предоставленном заказчиком. Датасет содержит комментарии пользоватетей, размеченные на позитивные и негативные.

Работа проводилась с помощью программного обеспечения JupyterHub средствами Python, библиотек pandas, numpy, nltk, sklearn, pymystem3, re, lightgbm. Использовались модели машинного обучения LogisticRegression(), RandomForestClassifier(), LGBMClassifier().

### Ход выполнения проекта

В ходе работы выполнено следующее:
- загрузка и предварительная обработка данных;
- лемматизация данных;
- расчёт TF-IDF, использование TF-IDF в качестве признаков для ML;
- разделение данных на стратифицированные обучающую и тестовую выборки;
- обучение, подбор гиперпараметров и кросс-валидация моделей LogisticRegression(), RandomForestClassifier(), LGBMClassifier(), SVC(), LinearSVC() с применением GridSearchCV(), расчёт F1-метрики при кросс-валидации;
- выбор лучшей модели и расчёт её F1-метрики на тестовой выборке.

### Результаты исследования

- Лучшей моделью признана SVC()

- Оптимальные гиперпараметры:
    - 'class_weight': 'balanced',
    - 'max_iter': -1 (не ограничено),
    - 'kernel': 'rbf'

- F1-метрика на тестовой выборке: 0.76

### Практические рекомендации

Можно рекомендовать использование SVC() с указанными гиперпараметрами для классификации комментариев на основе TF-IDF.

<div class="alert alert-block alert-success">
<b>Успех:</b> Приятно видеть подробный вывод в конце проекта!
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Данные загружены и подготовлены
- [ ]  Модели обучены
- [ ]  Значение метрики *F1* не меньше 0.75
- [ ]  Выводы написаны